src/pkg/services/m365/api/channels_pager.go

```diff
@@ -1,43 +1,223 @@
+		Client().
+		Teams().
+		ByTeamId(teamID).
+		Channels().
+		ByChannelId(channelID).
+		Messages().
+		Delta()
+
+	options := &teams.ItemChannelsItemMessagesDeltaRequestBuilderGetRequestConfiguration{
+		Headers: newPreferHeaders(preferPageSize(maxNonDeltaPageSize)),
+	}
+
+	return &channelMessageDeltaPageCtrl{
+		resourceID: teamID,
+		channelID:  channelID,
+		builder:    builder,
+		gs:         c.Stable,
+		options:    options,
+	}
+}
+
+// GetChannelMessagesDelta fetches a delta of all messages in the channel.
+func (c Channels) GetChannelMessagesDelta(
+	ctx context.Context,
+	teamID, channelID, prevDelta string,
+) ([]models.ChatMessageable, DeltaUpdate, error) {
+	var (
+		vs               = []models.ChatMessageable{}
+		pager            = c.NewChannelMessageDeltaPager(teamID, channelID)
+		invalidPrevDelta = len(prevDelta) == 0
+		newDeltaLink     string
+	)
+
+	// Loop through all pages returned by Graph API.
+	for {
+		page, err := pager.GetPage(graph.ConsumeNTokens(ctx, graph.SingleGetOrDeltaLC))
+		if graph.IsErrInvalidDelta(err) {
+			logger.Ctx(ctx).Infow("Invalid previous delta", "delta_link", prevDelta)
+
+			invalidPrevDelta = true
+
+			pager.Reset(ctx)
+
+			continue
+		}
+
+		if err != nil {
+			return nil, DeltaUpdate{}, graph.Wrap(ctx, err, "retrieving page of channels")
+		}
+
+		vals, err := pager.ValuesIn(page)
+		if err != nil {
+			return nil, DeltaUpdate{}, graph.Wrap(ctx, err, "extracting channels from response")
+		}
+
+		vs = append(vs, vals...)
+
+		nextLink, deltaLink := NextAndDeltaLink(page)
+
+		if len(deltaLink) > 0 {
+			newDeltaLink = deltaLink
+		}
+
+		if len(nextLink) == 0 {
+			break
+		}
 
-// type messagePageCtrl struct {
-// 	gs      graph.Servicer
-// 	builder *teams.ItemChannelsItemMessagesRequestBuilder
-// 	options *teams.ItemChannelsItemMessagesRequestBuilderGetRequestConfiguration
-// }
+		pager.SetNext(nextLink)
+	}
+
+	logger.Ctx(ctx).Debugf("retrieved %d channel messages", len(vs))
+
+	du := DeltaUpdate{
+		URL:   newDeltaLink,
+		Reset: invalidPrevDelta,
+	}
+
+	return vs, du, nil
+}
 
 // ---------------------------------------------------------------------------
 // channel pager
 // ---------------------------------------------------------------------------
 
-type ChannelDeltaEnumerator interface {
-	DeltaGetPager
-	ValuesInPageLinker[models.Channelable]
-	SetNextLinker
+var _ Pager[models.Channelable] = &channelPageCtrl{}
+
+type channelPageCtrl struct {
+	gs      graph.Servicer
+	builder *teams.ItemChannelsRequestBuilder
+	options *teams.ItemChannelsRequestBuilderGetRequestConfiguration
+}
+
+func (p *channelPageCtrl) SetNext(nextLink string) {
+	p.builder = teams.NewItemChannelsRequestBuilder(nextLink, p.gs.Adapter())
+}
+
+func (p *channelPageCtrl) GetPage(
+	ctx context.Context,
+) (PageLinker, error) {
+	resp, err := p.builder.Get(ctx, p.options)
+	return resp, graph.Stack(ctx, err).OrNil()
+}
+
+func (p *channelPageCtrl) GetOdataNextLink() *string {
+	return ptr.To("")
```

Comment:
is this intentionally returning "no next page"? If it is could you please add a comment that states why it's alright to assume there's no next page

Reply:
Hah, no, not intended.  This is copied code which slipped through my attention.  Good catch, ty!